experimental/zerotrie/src/reader.rs

```diff
@@ -0,0 +1,454 @@
+        // Note: We're trusting the compiler to inline this and remove the assertion
+        // hiding on the top of slice::split_at: `assert(mid <= self.len())`
+        Some(slice.split_at(mid))
+    }
+}
+
+#[inline]
+fn debug_get(slice: &[u8], index: usize) -> Option<u8> {
+    match slice.get(index) {
+        Some(x) => Some(*x),
+        None => {
+            debug_assert!(false, "debug_get: index expected to be in range");
+            None
+        }
+    }
+}
+
+#[inline]
+fn debug_get_range(slice: &[u8], range: Range<usize>) -> Option<&[u8]> {
+    match slice.get(range) {
+        Some(x) => Some(x),
+        None => {
+            debug_assert!(false, "debug_get_range: indices expected to be in range");
+            None
+        }
+    }
+}
+
+/// Given a slice starting with an offset table, returns the trie for the given index.
+///
+/// Arguments:
+/// - `trie` = a trie pointing at an offset table (after the branch node and search table)
+/// - `i` = the desired index within the offset table
+/// - `n` = the number of items in the offset table
+/// - `w` = the width of the offset table items minus one
+#[inline]
+fn get_branch(mut trie: &[u8], i: usize, n: usize, mut w: usize) -> Option<&[u8]> {
+    let mut p = 0usize;
+    let mut q = 0usize;
+    loop {
+        let indices;
+        (indices, trie) = debug_split_at(trie, n - 1)?;
+        p = (p << 8)
+            + if i == 0 {
+                0
+            } else {
+                debug_get(indices, i - 1)? as usize
+            };
+        q = match indices.get(i) {
+            Some(x) => (q << 8) + *x as usize,
+            None => trie.len(),
+        };
+        if w == 0 {
+            break;
+        }
+        w -= 1;
+    }
+    debug_get_range(trie, p..q)
+}
+
+/// Version of [`get_branch()`] specialized for the case `w == 0` for performance
+#[inline]
+fn get_branch_w0(mut trie: &[u8], i: usize, n: usize) -> Option<&[u8]> {
+    let indices;
+    (indices, trie) = debug_split_at(trie, n - 1)?;
+    let p = if i == 0 {
+        0
+    } else {
+        debug_get(indices, i - 1)? as usize
+    };
+    let q = match indices.get(i) {
+        Some(x) => *x as usize,
+        None => trie.len(),
+    };
+    debug_get_range(trie, p..q)
+}
+
+enum ByteType {
+    Ascii,
+    Span,
+    Value,
+    Match,
+}
+
+impl core::fmt::Debug for ByteType {
+    fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+        use ByteType::*;
+        f.write_str(match *self {
+            Ascii => "a",
+            Span => "s",
+            Value => "v",
+            Match => "m",
+        })
+    }
+}
+
+#[inline]
+fn byte_type(b: u8) -> ByteType {
+    match b & 0b11100000 {
+        0b10000000 => ByteType::Value,
+        0b10100000 => ByteType::Span,
+        0b11000000 => ByteType::Match,
+        0b11100000 => ByteType::Match,
+        _ => ByteType::Ascii,
+    }
+}
+
+// DISCUSS: This function is 7% faster *on aarch64* if we assert a max on w.
```

Comment:
seems fine, we can `cfg(target_arch)` if we really want to be nitty about it

Reply:
but also might be overkill and hard to continuously keep measuring for change

Reply:
I lean toward landing the three versions of the reader function and then maybe removing the redundant one when more people have a chance to reproduce the benchmarks